scanner/scanner.go

```diff
@@ -0,0 +1,185 @@
+	staticDB     *database.DB
+	staticClam   *clamav.ClamAV
+	staticLogger *logrus.Logger
+}
+
+// New returns a new Scanner with the given parameters.
+func New(ctx context.Context, db *database.DB, clam *clamav.ClamAV, logger *logrus.Logger) (*Scanner, error) {
+	if ctx == nil {
+		return nil, errors.New("invalid context provided")
+	}
+	if db == nil {
+		return nil, errors.New("invalid DB provided")
+	}
+	if clam == nil {
+		return nil, errors.New("invalid ClamAV instance provided")
+	}
+	if logger == nil {
+		return nil, errors.New("invalid logger provided")
+	}
+	return &Scanner{
+		staticCtx:    ctx,
+		staticDB:     db,
+		staticClam:   clam,
+		staticLogger: logger,
+	}, nil
+}
+
+// SweepAndScan sweeps the DB for new skylinks, locks them, scans them,
+// and updates their records in the DB. The method returns a boolean flag that
+// instructs the caller to sleep after this call when there are no records to
+// be scanned.
+func (s Scanner) SweepAndScan(abort chan bool) error {
+	sl, err := s.staticDB.SweepAndLock(s.staticCtx)
+	if err != nil {
+		if !errors.Contains(err, database.ErrNoDocumentsFound) {
+			s.staticLogger.Warnf("error while trying to lock a new record: %s", err)
+		}
+		return err
+	}
+	if sl.Skylink == "" {
+		s.staticLogger.Warnf("SweepAndLock returned a record with an empty skylink. Record hash: %s", hex.EncodeToString(sl.Hash[:]))
+		return errors.New("empty skylink")
+	}
+	inf, desc, size, scannedSize, err := s.staticClam.ScanSkylink(sl.Skylink, abort)
+	if err != nil {
+		// Scanning failed, log the error and unlock the record for another attempt.
+		s.staticLogger.Debugln(errors.AddContext(err, "scanning failed"))
+		sl.Status = database.SkylinkStatusNew
+		sl.Timestamp = time.Now().UTC()
+		err = s.staticDB.SkylinkSave(s.staticCtx, sl)
+		if err != nil {
+			s.staticLogger.Debugln(errors.AddContext(err, "unlocking a skylink failed"))
+		}
+		return err
+	}
+	// Sanity check: scannedSize vs size.
+	if scannedSize > size {
+		s.staticLogger.Warnf("Scanned size (%d bytes) is more than the content size (%d bytes) for skylink %s", scannedSize, size, sl.Skylink)
+	}
+	// Once the scanning is complete, we no longer need the skylink. We want to
+	// remove it and only keep its hash. We don't want our database to be an
+	// index of nasty files.
+	sl.Skylink = ""
+	sl.Status = database.SkylinkStatusComplete
+	sl.Infected = inf
+	sl.InfectionDescription = desc
+	sl.Size = size
+	sl.ScannedAllContent = scannedSize == size
+	sl.ScannedAllOffsets = false
+	sl.Timestamp = time.Now().UTC()
+	err = s.staticDB.SkylinkSave(s.staticCtx, sl)
+	if err != nil {
+		s.staticLogger.Debugln(errors.AddContext(err, "updating a skylink's status failed"))
+	}
+	return err
+}
+
+// Start launches a background task that periodically scans the database for
+// new skylink records and sends them for scanning.
+func (s Scanner) Start() {
+	// abort channel which interrupts the current scanning operation
+	abort := make(chan bool, 1)
```

Comment:
Why not get rid of this and pass `s.staticCtx` into `SweepAndScan` directly? Passing a context to interrupt an operation is very idiomatic. 

Also if you used a channel like that I'd close it instead of making it buffered and sending `true`. That's usually how you signal that an operation is done with a channel. Because then you can receive from that channel as many times as you want without getting blocked while sending a `true` will only allow 1 other thread to receive the value.

Reply:
I can make the channel within `SweepAndScan()` because that's what I need to pass to ClamAV. It seemed like a waste, though, to keep creating channels on each call when I can create one and keep it while the service was up.

Do you want me to create the channel within `SweepAndScan` and pass a context to it?

I agree about closing the channel - I'll make that change.

Reply:
From offline:
Since we need to pass a channel anyway, we decided that the current solution is acceptable.